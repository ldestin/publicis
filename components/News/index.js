import React, { useState, useMemo } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import Hits from '../Hits'
import Chart from '../Chart'
import { ORDERS } from '../../constants'

const orders = [ORDERS.TOP, ORDERS.NEW]

const News = ({ order: currentOrder, news, onPrevClick, onNextClick, onUpvoteClick }) => {
  const [hiddenHits, setHiddenHits] = useState({})
  const onHideClick = hit => {
    setHiddenHits({ ...hiddenHits, [hit.objectID]: true })
  }
  const currentHits = useMemo(() => {
    return news?.hits?.filter?.(hit => !hiddenHits[hit.objectID]) || []
  }, [news, hiddenHits])

  return (
    <div className="news">
      <div className="header">
        <img className="logo" alt="logo" src="/images/logo.gif" />
        <div className="links">
          {orders.reduce((acc, order, index) => {
            const isSelected = currentOrder === order
            const href = isSelected ? undefined : order

            acc.push(
              <a key={order} href={href} className={classnames('button link', order, { selected: isSelected })}>{order}</a> 
            )

            if (index < orders.length - 1) acc.push(' | ')

            return acc
          }, [])}
        </div>
      </div>
      {currentHits && <Hits hits={currentHits} maxPoints={news.maxPoints} onHideClick={onHideClick} onUpvoteClick={onUpvoteClick} />}
      <div className="buttons">
        {news && news.page > 0 && <div className="button prev-button" onClick={onPrevClick}>Prev</div>}
        {news && news.page < news.nbPages - 1 && <div className="button next-button" onClick={onNextClick}>Next</div>}
      </div>
      {currentHits && <Chart hits={currentHits} />}
    </div>
  )
}

News.defaultProps = {
  order: ORDERS.TOP,
}

News.propTypes = {
  order: PropTypes.oneOf(orders),
  onPrevClick: PropTypes.func,
  onNextClick: PropTypes.func,
  onUpvoteClick: PropTypes.func,
  news: PropTypes.shape({
    hits: PropTypes.arrayOf(PropTypes.shape({
      num_comments: PropTypes.number,
      points: PropTypes.number,
      title: PropTypes.string,
      url: PropTypes.string,
      author: PropTypes.string,
      created_at: PropTypes.string
    })),
    maxPoints: PropTypes.number,
    page: PropTypes.number,
    nbPages: PropTypes.number,
  })
}

export default News
