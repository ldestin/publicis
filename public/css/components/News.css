.news {
  background: var(--grey-bg);
  width: 85%;
  margin: auto;
}

.news .header {
  display: flex;
  background: var(--orange);
  padding: 6px 4px;
}

.news .logo {
  border: 1px solid var(--white);
}

.news .links {
  margin-left: 1rem;
}

.news .link.selected {
  color: var(--white);
}

.news .buttons {
  padding: 1rem 0;
}

@media (max-width: 700px) {
  .news {
    width: 100%;
    margin: 0;
  }
}
