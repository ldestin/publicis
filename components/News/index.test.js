import { describe, expect, jest } from '@jest/globals'
import { render } from '@testing-library/react'
import React from 'react'
import News from '.'

jest.mock('../Hits', () => props => <div>Hits: {JSON.stringify(props, null, 2)}</div>)
jest.mock('../Chart', () => props => <div>Chart: {JSON.stringify(props, null, 2)}</div>)

const props = {
  news: {
    hits: [{}],
    maxPoints: 10,
    page: 3,
    nbPages: 5,
    onPrevClick: jest.fn(),
    onNextClick: jest.fn(),
    onUpvoteClick: jest.fn(),
  }
}

it('should render correctly', () => {
  const { container } = render(<News {...props} />)

  expect(container).toMatchSnapshot()
  expect(container.querySelector('.prev-button')).toBeTruthy()
  expect(container.querySelector('.next-button')).toBeTruthy()
})

it('should render correctly without hits', () => {
  const { container } = render(<News {...props} news={{ hits: undefined }} />)

  expect(container).toMatchSnapshot()
  expect(container.querySelector('.prev-button')).toBeFalsy()
  expect(container.querySelector('.next-button')).toBeFalsy()
})

it('should render correctly on first page', () => {
  const { container } = render(<News {...props} news={{ ...props.news, page: 0 }} />)

  expect(container).toMatchSnapshot()
  expect(container.querySelector('.prev-button')).toBeFalsy()
  expect(container.querySelector('.next-button')).toBeTruthy()
})

it('should render correctly on last page', () => {
  const { container } = render(<News {...props} news={{ ...props.news, page: 4 }} />)

  expect(container).toMatchSnapshot()
  expect(container.querySelector('.prev-button')).toBeTruthy()
  expect(container.querySelector('.next-button')).toBeFalsy()
})
