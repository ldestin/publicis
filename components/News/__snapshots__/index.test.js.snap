// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="news"
  >
    <div
      id="api"
    />
    <div
      class="header"
    >
      <img
        alt="logo"
        class="logo"
        src="/images/logo.gif"
      />
      <div
        class="links"
      >
        <a
          class="button link top selected"
        >
          top
        </a>
         | 
        <a
          class="button link new"
          href="new"
        >
          new
        </a>
      </div>
    </div>
    <div>
      Hits: 
      {
  "hits": [
    {}
  ],
  "maxPoints": 10
}
    </div>
    <div
      class="buttons"
    >
      <div
        class="button prev-button"
      >
        Prev
      </div>
      <div
        class="button next-button"
      >
        Next
      </div>
    </div>
    <div>
      Chart: 
      {
  "hits": [
    {}
  ]
}
    </div>
  </div>
</div>
`;

exports[`should render correctly on first page 1`] = `
<div>
  <div
    class="news"
  >
    <div
      id="api"
    />
    <div
      class="header"
    >
      <img
        alt="logo"
        class="logo"
        src="/images/logo.gif"
      />
      <div
        class="links"
      >
        <a
          class="button link top selected"
        >
          top
        </a>
         | 
        <a
          class="button link new"
          href="new"
        >
          new
        </a>
      </div>
    </div>
    <div>
      Hits: 
      {
  "hits": [
    {}
  ],
  "maxPoints": 10
}
    </div>
    <div
      class="buttons"
    >
      <div
        class="button next-button"
      >
        Next
      </div>
    </div>
    <div>
      Chart: 
      {
  "hits": [
    {}
  ]
}
    </div>
  </div>
</div>
`;

exports[`should render correctly on last page 1`] = `
<div>
  <div
    class="news"
  >
    <div
      id="api"
    />
    <div
      class="header"
    >
      <img
        alt="logo"
        class="logo"
        src="/images/logo.gif"
      />
      <div
        class="links"
      >
        <a
          class="button link top selected"
        >
          top
        </a>
         | 
        <a
          class="button link new"
          href="new"
        >
          new
        </a>
      </div>
    </div>
    <div>
      Hits: 
      {
  "hits": [
    {}
  ],
  "maxPoints": 10
}
    </div>
    <div
      class="buttons"
    >
      <div
        class="button prev-button"
      >
        Prev
      </div>
    </div>
    <div>
      Chart: 
      {
  "hits": [
    {}
  ]
}
    </div>
  </div>
</div>
`;

exports[`should render correctly without hits 1`] = `
<div>
  <div
    class="news"
  >
    <div
      id="api"
    />
    <div
      class="header"
    >
      <img
        alt="logo"
        class="logo"
        src="/images/logo.gif"
      />
      <div
        class="links"
      >
        <a
          class="button link top selected"
        >
          top
        </a>
         | 
        <a
          class="button link new"
          href="new"
        >
          new
        </a>
      </div>
    </div>
    <div>
      Hits: 
      {
  "hits": []
}
    </div>
    <div
      class="buttons"
    />
    <div>
      Chart: 
      {
  "hits": []
}
    </div>
  </div>
</div>
`;
